core/services/pipeline/orm.go

```diff
@@ -0,0 +1,360 @@
+
+		// Update the task run record with the output and error
+		if result.Value != nil {
+			out := &JSONSerializable{Val: result.Value}
+			err = tx.Exec(`UPDATE pipeline_task_runs SET output = ?, error = NULL, finished_at = ? WHERE id = ?`, out, time.Now(), ptRun.ID).Error
+		} else if result.Error != nil {
+			logger.Errorw("Error in pipeline task", "error", result.Error)
+			errString := null.StringFrom(result.Error.Error())
+			err = tx.Exec(`UPDATE pipeline_task_runs SET output = NULL, error = ?, finished_at = ? WHERE id = ?`, errString, time.Now(), ptRun.ID).Error
+		}
+		if err != nil {
+			return errors.Wrap(err, "could not mark pipeline_task_run as finished")
+		}
+		return nil
+	})
+	return done, err
+}
+
+const (
+	postgresChannelRunStarted   = "pipeline_run_started"
+	postgresChannelRunCompleted = "pipeline_run_completed"
+)
+
+func (o *orm) ListenForNewRuns() (*utils.PostgresEventListener, error) {
+	listener := &utils.PostgresEventListener{
+		URI:                  o.uri,
+		Event:                postgresChannelRunStarted,
+		MinReconnectInterval: 1 * time.Second,
+		MaxReconnectDuration: 1 * time.Minute,
+	}
+	err := listener.Start()
+	if err != nil {
+		return nil, err
+	}
+	return listener, nil
+}
+
+// AwaitRun waits until a run has completed (either successfully or with errors)
+// and then returns.  It uses two distinct methods to determine when a job run
+// has completed:
+//    1) periodic polling
+//    2) Postgres notifications
+func (o *orm) AwaitRun(ctx context.Context, runID int64) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// This goroutine polls the DB at a set interval
+	chPoll := make(chan error)
+	go func() {
+		var err error
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			var done bool
+			done, err = runFinished(o.db, runID)
+			if err != nil || done {
+				break
+			}
+			time.Sleep(1 * time.Second)
+		}
+
+		select {
+		case chPoll <- err:
+		case <-ctx.Done():
+		}
+	}()
+
+	// This listener subscribes to the Postgres event informing us of a completed pipeline run
+	listener := utils.PostgresEventListener{
+		URI:                  o.uri,
+		Event:                postgresChannelRunCompleted,
+		PayloadFilter:        fmt.Sprintf("%v", runID),
+		MinReconnectInterval: 1 * time.Second,
+		MaxReconnectDuration: 1 * time.Minute,
+	}
+	err := listener.Start()
+	if err != nil {
+		return err
+	}
+	defer listener.Stop()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-chPoll:
+		return err
+	case <-listener.Events():
+		return nil
+	}
+}
+
+func (o *orm) ResultsForRun(runID int64) ([]Result, error) {
+	var results []Result
+	err := utils.GormTransaction(o.db, func(tx *gorm.DB) (err error) {
+		done, err := runFinished(tx, runID)
+		if err != nil {
+			return err
+		} else if !done {
+			return errors.New("can't fetch run results, run is still in progress")
+		}
+
+		var taskRuns []TaskRun
+		err = o.db.
```

Comment:
runFinished probably not necessary, just move the filter into this query?

Reply:
It's used in 2 methods